gpflow/training/monitor/monitor.py

```diff
@@ -0,0 +1,790 @@
+    .with_name('print')\
+    .with_condition(mon.PeriodicIterationCondition(10))\
+
+# This task will create a Tensoreflow summary of the model for Tensorboard. It will run at
+# every 100-th iteration. We also want to do this after the optimisation is finished.
+#
+tensorboard_task = mon.StandardTensorBoardTask(model, './model-tensorboard')\
+    .with_name('tensorboard')\
+    .with_condition(mon.PeriodicIterationCondition(100))\
+    .with_exit_condition(True)
+
+# This task will save the Tensorflow session every 15-th iteration.
+# The directory pointed to by `checkpoint_dir` must exist.
+#
+checkpoint_task = mon.CheckpointTask(checkpoint_dir="./model-saves")\
+    .with_name('checkpoint')\
+    .with_condition(mon.PeriodicIterationCondition(15))\
+
+monitor_tasks = [print_task, tensorboard_task, checkpoint_task]
+
+# Create and start the monitor.
+#
+monitor = mon.Monitor(monitor_tasks, session, global_step)
+monitor.start_monitoring()
+
+try:
+    # Run the optimiser providing it with the monitor as the callback function
+    #
+    optimiser = gpflow.train.AdamOptimizer(0.01)
+    optimiser.minimize(model, step_callback=monitor, global_step=global_step)
+finally:
+    # Let the monitor know that the optimisation has finished and print the timing summary
+    #
+    monitor.stop_monitoring()
+    monitor.print_summary()
+"""
+
+import time
+import abc
+from typing import Callable, List, Dict, Optional, Iterator, Any
+import itertools
+from pathlib import PurePath
+
+import numpy as np
+import tensorflow as tf
+
+from ... import params_as_tensors_for
+from ... import settings
+from ...models import Model
+from ...params import Parameter
+from timeit import default_timer as timer
+
+
+def get_hr_time() -> float:
+    """
+    Gets high resolution time. Mainly defined here for the convenience of unit testing.
+    """
+    return timer()
+
+
+def create_global_step(tf_session: tf.Session) -> tf.Variable:
+    """
+    Creates the Tensorflow 'global_step' variable (see `MonitorContext.global_step_tensor`).
+    :param tf_session: Tensorflow session the optimiser is running in
+    :return: The variable tensor.
+    """
+    global_step_tensor = tf.Variable(0, trainable=False, name="global_step")
+    tf_session.run(global_step_tensor.initializer)
+    return global_step_tensor
+
+
+def restore_session(tf_session: tf.Session, checkpoint_dir: str,
+                    saver: Optional[tf.train.Saver] = None) -> None:
+    """
+    Restores Tensorflow session from the latest checkpoint.
+    :param tf_session: The TF session
+    :param checkpoint_dir: checkpoint files directory.
+    :param saver: The saver object, if not provided a default saver object will be created.
+    """
+    checkpoint_path = tf.train.latest_checkpoint(checkpoint_dir)
+    print("Restoring session from `%s`." % checkpoint_path)
+    saver = saver or get_default_saver()
+    saver.restore(tf_session, checkpoint_path)
+
+
+def get_default_saver() -> tf.train.Saver:
```

Comment:
Recommended: Add default argument for `max_to_keep`?

Reply:
OK, makes sense, done.